io/gdal/GDALReader.cpp

```diff
@@ -0,0 +1,216 @@
+{
+
+static PluginInfo const s_info = PluginInfo(
+        "readers.gdal",
+        "Read GDAL rasters as point clouds.",
+        "http://pdal.io/stages/reader.gdal.html");
+
+
+CREATE_STATIC_PLUGIN(1, 0, GDALReader, Reader, s_info);
+
+
+std::string GDALReader::getName() const
+{
+    return s_info.name;
+}
+
+
+GDALReader::GDALReader()
+    : m_index(0)
+{}
+
+
+void GDALReader::initialize()
+{
+    GlobalEnvironment::get().initializeGDAL(log());
+    if (!m_raster)
+        m_raster = std::unique_ptr<gdal::Raster>(new gdal::Raster(m_filename));
+
+    m_raster->open();
+    setSpatialReference(m_raster->getSpatialRef());
+    m_count = m_raster->m_raster_x_size * m_raster->m_raster_y_size;
+
+}
+
+void GDALReader::processOptions(const Options& options)
+{
+
+}
+
+QuickInfo GDALReader::inspect()
+{
+    QuickInfo qi;
+    std::unique_ptr<PointLayout> layout(new PointLayout());
+
+    addDimensions(layout.get());
+    initialize();
+
+    m_raster = std::unique_ptr<gdal::Raster>(new gdal::Raster(m_filename));
+    m_raster->open();
+
+//     Dimension::IdList dims = layout->dims();
+//     for (auto di = dims.begin(); di != dims.end(); ++di)
+//         qi.m_dimNames.push_back(layout->dimName(*di));
+//     if (!Utils::numericCast(m_lasHeader.pointCount(), qi.m_pointCount))
+    qi.m_pointCount = m_raster->m_raster_x_size * m_raster->m_raster_y_size;
+//     qi.m_bounds = m_lasHeader.getBounds();
+    qi.m_srs = m_raster->getSpatialRef();
+    qi.m_valid = true;
+
+    PointTable table;
+    done(table);
+
+    return qi;
+}
+
+
+void GDALReader::addDimensions(PointLayoutPtr layout)
+{
+    int nBands = m_raster->m_band_count;
+
+    layout->registerDim(pdal::Dimension::Id::X);
+    layout->registerDim(pdal::Dimension::Id::Y);
+    for (int i=0; i < nBands; ++i)
+    {
+        std::ostringstream oss;
+        oss << "band-" << i+1;
+        layout->registerOrAssignDim(oss.str(), Dimension::Type::Double);
+    }
+}
+
+
+void GDALReader::ready(PointTableRef table)
+{
+    m_index = 0;
+
+
+}
+
+template<typename T>
+double convert(uint8_t*p)
+{
+    double output;
+
+    T t;
+    std::copy(p, p + sizeof(t), (uint8_t*)&t);
+    output = static_cast<double>(t);
+    return output;
+}
+
+point_count_t GDALReader::read(PointViewPtr view, point_count_t num)
+{
+    point_count_t count = std::min(count, m_count - m_index);
+    PointId nextId = view->size();
+
+    std::array<double, 2> coords;
+    for (int row = 0; row < m_raster->m_raster_y_size; ++row)
+    {
+        for(int col = 0; col < m_raster->m_raster_x_size; ++col)
+        {
+            m_raster->pixelToCoord(row, col, coords);
+            view->setField(Dimension::Id::X, nextId, coords[0]);
+            view->setField(Dimension::Id::Y, nextId, coords[1]);
+            nextId++;
+        }
+    }
+
+    std::vector<uint8_t> band;
+    std::vector<pdal::Dimension::Type::Enum> band_types = m_raster->getPDALDimensionTypes();
+    for (int b = 0; b < m_raster->m_band_count; ++b)
```

Comment:
GDAL indexes bands from 1 and everything else from 0. It seemed safer to use 0 and let GDAL tell me there's no "band-0" instead of doing alist[-1] and blowing up.